### PR TITLE
RestoringGlobalsComponent: Store value on shutdown

### PR DESCRIPTION
**Pull request by RoboMagus**
# What does this implement/fix?

This change will ensure the last value of a `RestoringGlobalsComponent` is stored to flash before shutdown is called. For example when in a lambda a value is stored right before `App.reboot();` is called.

Requires PR #3585 to work propperly!

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
...

globals:
  - id: failure_restart_count
    type: uint32_t
    initial_value: '0'
    restore_value: yes

logger:
  on_message:
    level: WARN
    then:
      - lambda: |-
          static uint8_t failure_count = 0;
          if(strstr(message, "cirtical failure")) {
            failure_count++;
          }
          if (failure_count >= 5) {
            id(failure_restart_count) += 1;
            delay(250);  // NOLINT
            App.safe_reboot();
          }
...
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/globals/globals_component.h [hunk 1]
```diff
@@ -45,16 +45,24 @@ template<typename T> class RestoringGlobalsComponent : public Component {
...
+  }
+
+  void on_shutdown() override {
+    store_value();
+  }
+
+  void set_name_hash(uint32_t name_hash) { this->name_hash_ = name_hash; }
+
+ protected:
+  void store_value() {
```
Comment (OttoWinter) [inline esphome/components/globals/globals_component.h]:
```suggestion
  void store_value_() {
```
Needed to match the naming conventions

File: esphome/components/globals/globals_component.h [hunk 2]
```diff
@@ -45,16 +45,24 @@ template<typename T> class RestoringGlobalsComponent : public Component {
   float get_setup_priority() const override { return setup_priority::HARDWARE; }
 
   void loop() override {
+    store_value();
+  }
+
+  void on_shutdown() override {
+    store_value();
```
Comment (OttoWinter) [inline esphome/components/globals/globals_component.h]:
```suggestion
    store_value_();
```

File: esphome/components/globals/globals_component.h [hunk 3]
```diff
@@ -45,16 +45,24 @@ template<typename T> class RestoringGlobalsComponent : public Component {
   float get_setup_priority() const override { return setup_priority::HARDWARE; }
 
   void loop() override {
+    store_value();
```
Comment (OttoWinter) [inline esphome/components/globals/globals_component.h]:
```suggestion
    store_value_();
```

**Review by OttoWinter — APPROVED** [review]
Thanks!

**Comment by RoboMagus** [comment]
Thanks @OttoWinter for taking a look. 
I've updated the styling as per your suggestion when making `clang-format` happy again.

I'd suggest *not* merging this PR before #3585 though, as that one is sort of a prerequisite ; )
(As currently the preferences would have it's last sync to flash before globals are shut down)

**Review by jesserockz — APPROVED** [review]

**Pull request merged**